src/CoreWCF.Primitives/src/CoreWCF/Security/NegotiationTokenAuthenticator.cs

```diff
@@ -0,0 +1,1056 @@
+        }
+
+        // helpers
+        protected SecurityContextSecurityToken IssueSecurityContextToken(UniqueId contextId, string id, byte[] key,
+            DateTime tokenEffectiveTime, DateTime tokenExpirationTime,
+            ReadOnlyCollection<IAuthorizationPolicy> authorizationPolicies, bool isCookieMode)
+        {
+            return IssueSecurityContextToken(contextId, id, key, tokenEffectiveTime, tokenExpirationTime, null,
+                tokenEffectiveTime, tokenExpirationTime, authorizationPolicies, isCookieMode);
+        }
+
+        protected SecurityContextSecurityToken IssueSecurityContextToken(UniqueId contextId, string id, byte[] key,
+            DateTime tokenEffectiveTime, DateTime tokenExpirationTime, UniqueId keyGeneration, DateTime keyEffectiveTime,
+            DateTime keyExpirationTime, ReadOnlyCollection<IAuthorizationPolicy> authorizationPolicies, bool isCookieMode)
+        {
+          //  this.CommunicationObject.ThrowIfClosedOrNotOpen();
+            if (this.securityStateEncoder == null && isCookieMode)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.SctCookieNotSupported)));
+            }
+            byte[] cookieBlob = (isCookieMode) ? this.cookieSerializer.CreateCookieFromSecurityContext(contextId, id, key, tokenEffectiveTime, tokenExpirationTime, keyGeneration,
+                                keyEffectiveTime, keyExpirationTime, authorizationPolicies) : null;
+
+            SecurityContextSecurityToken issuedToken = new SecurityContextSecurityToken(contextId, id, key, tokenEffectiveTime, tokenExpirationTime,
+                authorizationPolicies, isCookieMode, cookieBlob, keyGeneration, keyEffectiveTime, keyExpirationTime);
+            return issuedToken;
+        }
+
+        void InitializeDefaults()
+        {
+            this.encryptStateInServiceToken = !defaultServerMaintainState;
+            this.serviceTokenLifetime = defaultServerIssuedTokenLifetime;
+            this.maximumCachedNegotiationState = defaultServerMaxActiveNegotiations;
+            this.negotiationTimeout = defaultServerMaxNegotiationLifetime;
+            this.isClientAnonymous = false;
+            this.standardsManager = defaultStandardsManager;
+            this.securityStateEncoder = defaultSecurityStateEncoder;
+            this.maximumConcurrentNegotiations = defaultServerMaxActiveNegotiations;
+            // we rely on the transport encoders to enforce the message size except in the 
+            // mixed mode nego case, where the client is unauthenticated and the maxMessageSize is too
+            // large to be a mitigation
+            this.maxMessageSize = Int32.MaxValue;
+        }
+
+        public override Task CloseAsync(CancellationToken token)
+        {
+            if (this.negotiationHost != null)
+            {
+               // this.negotiationHost.CloseAsync(token);
+                this.negotiationHost = null;
```

Comment:
Does this need to be disposed?

Reply:
not needed... no unmanaged resources... LMK what you think ?

Reply:
As no managed resources, can you remove the commented out CloseAsync

---
In reply to: [565709230](https://github.com/CoreWCF/CoreWCF/pull/279#discussion_r565709230) [](ancestors = 565709230)